provider/image.go

```diff
@@ -0,0 +1,314 @@
+package provider
+
+import (
+	"bufio"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/jsonmessage"
+	structpb "github.com/golang/protobuf/ptypes/struct"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+)
+
+const defaultDockerfile = "Dockerfile"
+
+type Image struct {
+	Name     string
+	SkipPush bool
+	Registry Registry
+	Tag      string
+	Build    Build
+}
+
+type Registry struct {
+	Server   string
+	Username string
+	Password string
+}
+
+type Build struct {
+	Context      string
+	Dockerfile   string
+	CachedImages []string
+	Env          map[string]string
+	Args         map[string]*string
+	ExtraOptions []string
+	Target       string
+}
+
+func (p *dockerNativeProvider) dockerBuild(ctx context.Context,
+	urn resource.URN,
+	props *structpb.Struct) (string, *structpb.Struct, error) {
+
+	inputs, err := plugin.UnmarshalProperties(props, plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return "", nil, err
+	}
+
+	//set Registry
+	var reg Registry
+	if !inputs["registry"].IsNull() {
+		reg = Registry{
+			Server:   inputs["registry"].ObjectValue()["server"].StringValue(),
```

Comment:
Interesting! This could be slightly cleaner if the marshaling was moved to a separate function, but I think what's hiding here is a feature request for platform to enable an SDK for hydrating typed structs out of PropertyMap and taking care of all the nasty corner cases, similar to how Terraform Plugin framework does it.

Reply:
I can certainly separate this out, since I want to add test cases in a followup anyhow.